### PR TITLE
Fix base URL issue/cookie name issue

### PR DESCRIPTION
**Pull request by hosainnet**
# Description

Our pipeline has been broken because intermittently the IDAM redirect_url/cookie hostname are incorrect. These shared values were being overwitten on every request to the homepage based on the request's hostname (causing a race condition).
This fix ensures that redirectUri/cookie hostName are scoped per request and don't modify the original idamArgs.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/idam.js [hunk 1]
```diff
@@ -22,11 +24,6 @@ const methods = {
   getIdamArgs: () => {
     return idamArgs;
   },
-  setRedirectUri: (req, res, next) => {
-    idamArgs.hostName = req.get('host').split(':')[0];
-    idamArgs.redirectUri = `https://${req.get('host')}${config.paths.authenticated}`;
```
Comment (mthava) [inline services/idam.js]: You may have to keep this line. When you go live (public). When you go live you want the redirect to be forwarded to public DNS not the internal. With the new changes when idam tries to redirects (302 - browser redirect) the request to internal URL will fail.

File: test/unit/services/idam.test.js [hunk 2]
```diff
@@ -85,4 +85,56 @@ describe(modulePath, () => {
...
+  describe('idam args', () => {
+    let authenticated = '';
+    let baseUrl = '';
+    let indexUrl = '';
+    let idamApiUrl = '';
+    let idamLoginUrl = '';
+    let idamSecret = '';
+    let idamClientID = '';
+
+    before(() => {
```
Comment (alexwakeman) [inline test/unit/services/idam.test.js]: All looks good - just finding this unit test a bit convoluted - you could use a `require` override util like [mock-require](https://www.npmjs.com/package/mock-require) to mock the config and avoid affecting other unit tests.

File: services/idam.js (hunk 1 above)
Comment (hosainnet) [reply services/idam.js]: Good point, update the code to do this per request but not modify the global state

File: test/unit/services/idam.test.js (hunk 2 above)
Comment (hosainnet) [reply test/unit/services/idam.test.js]: Sorted using Sinon sandbox 👍

**Comment by hosainnet** [comment]
@dtaylor7 any thoughts? As I'll be doing something similar for DN

File: services/idam.js [hunk 3]
```diff
@@ -22,13 +23,13 @@ const methods = {
...
-    next();
-  },
-  authenticate: (...args) => {
-    return middleware.authenticate(idamArgs, ...args);
+  authenticate: (req, res, next) => {
+    const reqUrl = url.parse(req.get('host'));
+    // clone args so we don't modify the global idamArgs
+    const args = Object.assign({}, idamArgs);
+    args.hostName = reqUrl.hostname;
+    args.redirectUri = `${reqUrl.protocol}//${reqUrl.host + config.paths.authenticated}`;
```
Comment (mthava) [inline services/idam.js]: 👍

**Comment by dtaylor7** [comment]
Looks good, not sure why it would the request would be wrong when navigating to start page but i'm sure it does. Think i prefer this was of doing it anyway.

**Comment by hosainnet** [comment]
Replaced with #83

**Pull request closed without merging**